src/app/doc_api.cpp

```diff
@@ -261,17 +261,46 @@ bool DocApi::cropCel(LayerImage* layer,
       newCelBounds.offset(-bounds.origin());
 
       gfx::Point paintPos(newCelBounds.x - newCelPos.x,
-                          newCelBounds.y - newCelPos.y);
+        newCelBounds.y - newCelPos.y);
 
+      const color_t bg = image->pixelFormat() == IMAGE_TILEMAP ?
+                           notile :
+                           m_document->bgColor(layer);
       newCelPos = newCelBounds.origin();
 
-      // Crop the image
+      doc::Grid grid;
+      if (layer->isTilemap()) {
+        const Tileset* tileset = static_cast<LayerTilemap*>(layer)->tileset();
+        grid = tileset->grid();
+
+        const gfx::Point roiOrigin = bounds.origin() - cel->position();
+        const gfx::Rect  roiRect(roiOrigin, bounds.size());
+        const gfx::Size  imageTileSize = grid.canvasToTile(roiRect).size();
+        const gfx::Point newPaintPos = grid.canvasToTile(paintPos);
+        const gfx::Point celPaintPixelPos = grid.tileToCanvas(newPaintPos);
+        const gfx::Point celPosDifference = paintPos - celPaintPixelPos;
+        const gfx::Size  imageSizeInPixel = grid.tilemapSizeToCanvas(imageTileSize);
+
+        if (imageSizeInPixel.w >= newCelBounds.w)
+          newCelPos.x -= celPosDifference.x;
+        else
+          newCelPos.x += celPosDifference.x;
+
+        if (imageSizeInPixel.h >= newCelBounds.h)
+          newCelPos.y -= celPosDifference.y;
+        else
+          newCelPos.y += celPosDifference.y;
+
+        paintPos = newPaintPos;
+        newCelBounds.setSize(imageTileSize);
```

Comment:
This can be simplified. Position arithmetics and some CanvasToTiles conversion is enough.
Take care of this variables:
1. `paintPos` will be the start coordinate in tile units refered to the tilemap image (for example: paintPos = (0, 0) is the most left-top tile of the cel image). It's used in `crop_image` function.
2. `newCelBounds` is used to define the crop size (in tile units) of `crop_image` function.
3. `newCelPos` it'll the "final" cel position in pixels, but be careful,  in the following `shrink`, `newCelPos` may need some adjust.

Reply:
Thank you for the review. After taking another look at the code, I realized the value of the below arithmetic would always be equal to `paintPos` and thus makes `roiOringin` unnecessary. The only single occurrence of it can be substituted with `paintPos`
```
const gfx::Point roiOrigin = bounds.origin() - cel->position();
const gfx::Rect  roiRect(roiOrigin, bounds.size());
```
to give this.
```
const gfx::Rect  roiRect(paintPos, bounds.size());
```
However, I think the other calculations are necessary so I'll walk you through my thought process. Perhaps in trying to explain it, you and/or I can come up with a better alternative to make it more elegant than it is now.

`imageTileSize` and `newPaintPos`, as you mentioned, are quite straightforward -- they're the conversion of the image/coordinate from pixel to tile unit and they're used as-is on lines 294 and 295 respectively. We, however, need to get the position of the `cel` right, otherwise the positioning of the image view would be incorrect. 

The result of using the point designated by `newPaintPos` on an image gives the entire part of the tile at that index. However, the bounds origin specified may have touched only a small part of the tile at that starting image index (or call it coordinate/point). So we needed a way to get how far out or in, from that tile we need to "go". Thus, we convert the `newPaintPos` back to pixel and find the difference between the "original" position specified in the bounds and the calculated position returned in `newPaintPos`. This is what I was trying to accomplish here.
```
celPaintPixelPos = grid.tileToCanvas(newPaintPos); // <-- the origin of the tile (in pixel) starting that image
celPosDifference = paintPos - celPaintPixelPos; // <- the difference between the start and the paintPos
```

Then, we adjusted the `celPosition`'s `x` and `y` coordinates based on the newly created image size.

Reply:
I see this:
![1632491476919](https://user-images.githubusercontent.com/32165475/134686511-59e6161b-14a8-4094-91d5-6c1ca1a992a6.jpg)